build-tools/src/main/kotlin/org/jetbrains/kotlin/benchmark/SwiftBenchmarkingPlugin.kt

```diff
@@ -0,0 +1,226 @@
+    }
+}
+
+/**
+ * A plugin configuring a benchmark Kotlin/Native project.
+ */
+open class SwiftBenchmarkingPlugin: Plugin<Project> {
+    private fun Project.determinePreset(): KotlinNativeTargetPreset = kotlin.presets.macosX64 as KotlinNativeTargetPreset
+
+    private fun Project.configureSourceSets(kotlinVersion: String) {
+        with(kotlin.sourceSets) {
+            commonMain.dependencies {
+                implementation("org.jetbrains.kotlin:kotlin-stdlib-common:$kotlinStdlibVersion")
+            }
+
+            project.configurations.getByName(nativeMain.implementationConfigurationName).apply {
+                // Exclude dependencies already included into K/N distribution (aka endorsed libraries).
+                exclude(mapOf("module" to "kotlinx.cli"))
+            }
+
+            repositories.maven {
+                it.setUrl(kotlinStdlibRepo)
+            }
+
+            // Add sources specified by a user in the benchmark DSL.
+            afterEvaluate {
+                benchmark.let {
+                    commonMain.kotlin.srcDirs(*it.commonSrcDirs.toTypedArray())
+                    nativeMain.kotlin.srcDirs(*(it.nativeSrcDirs).toTypedArray())
+                }
+            }
+        }
+    }
+
+    private fun Project.configureNativeTarget(hostPreset: KotlinNativeTargetPreset) {
+        kotlin.targetFromPreset(hostPreset, NATIVE_TARGET_NAME) {
+            compilations.getByName("main").kotlinOptions.freeCompilerArgs = project.compilerArgs + listOf("-l", "kotlinx-cli")
+            binaries.framework(NATIVE_FRAMEWORK_NAME, listOf(RELEASE)) {
+                // Specify settings configured by a user in the benchmark extension.
+                afterEvaluate {
+                    linkerOpts.addAll(benchmark.linkerOpts)
+                }
+            }
+        }
+    }
+
+    private fun Project.configureMPPExtension() {
+        configureSourceSets(kotlinVersion)
+        configureNativeTarget(determinePreset())
+    }
+
+    private fun Project.configureTasks() {
+        val nativeTarget = kotlin.targets.getByName(NATIVE_TARGET_NAME) as KotlinNativeTarget
+        val executable = Paths.get(buildDir.absolutePath, benchmark.applicationName)
+        // Build executable from swift code.
+        val framework = nativeTarget.binaries.getFramework(NATIVE_FRAMEWORK_NAME, NativeBuildType.RELEASE)
+        val buildSwift = tasks.create("buildSwift") { task ->
+            task.dependsOn(framework.linkTaskName)
+            task.doLast {
+                val frameworkParentDirPath = framework.outputDirectory.absolutePath
+                val options = listOf("-Xlinker", "-rpath", "-Xlinker", frameworkParentDirPath, "-F", frameworkParentDirPath)
+                compileSwift(project, nativeTarget.konanTarget, benchmark.swiftSources, options,
+                        Paths.get(buildDir.absolutePath, benchmark.applicationName), false)
+            }
+        }
+        framework.linkTask.finalizedBy(buildSwift)
```

Comment:
We have already added a dependency `buildSwift -> linkTask`. Why do we also need  this `finalizedBy` call?

Reply:
Because I want to have guarantee that at the end of link I'll have executable.

Reply:
> at the end of link I'll have executable.
How/where do we start the link process? As far as I understand we never run the framework link task directly. Instead, we run `konanRun` which depends on `buildSwift`. This is true?

Reply:
We mostly yes, but calling link task isn't prohibited + I have an idea that may be we can have one more only compilation mode for all benchmarks later so I wanted to have same behaviour for all subprojects. Can this both-side dependency cause some problems?

Reply:
> but calling link task isn't prohibited

Yes, but when a user calls the link task, he expects the framework to be linked and have no expectations about the swift executable. Or he even may not want to link the swift executable. 

> Can this both-side dependency cause some problems?

No, I just wanted to minimize unobvious relations between tasks.

> I wanted to have same behaviour for all subprojects.

I would prefer a separate task, say `buildBenchmark`, that depends on `linkExecutable...` or `buildSwift` tasks in different benchmarks. The `finalizedBy` will work too but it's less obvious.

Also the link tasks generated by the MPP plugin have different names: `linkExecutable...` for executables and `linkFramework` for frameworks, so we still won't be able to refer them uniformly.